### PR TITLE
Add storage size component to weights

### PR DESCRIPTION
**Pull request by KiChjang**
This PR adds the long-awaited storage size weight component to the `Weight` struct in `sp-weights`.

Part of #12176.

### CODE REVIEW

File: primitives/weights/src/weight_v2.rs
Comment (kianenigma) [inline primitives/weights/src/weight_v2.rs:44]: didn't we agree on `proof_size`?

File: primitives/weights/src/weight_v2.rs
Comment (kianenigma) [inline primitives/weights/src/weight_v2.rs:39]: This will need a migration?

File: primitives/weights/src/weight_v2.rs
Comment (ggwpez) [inline primitives/weights/src/weight_v2.rs:101]:
```suggestion
		if total.any_gt(limit) {
```
future proof

File: primitives/weights/src/weight_v2.rs
Comment (ggwpez) [inline primitives/weights/src/weight_v2.rs:129]: Just an idea: Could have opted for a private field iter, such that this becomes something like:

```suggestion
		self.fields_iter().map(|s| saturating_add(s, rhs)).collect::<Self>()
```

Then it is future proof when changing or adding fields. But could be overkill.

File: primitives/weights/src/weight_v2.rs
Comment (ggwpez) [inline primitives/weights/src/weight_v2.rs:300]: Not sure if we could make this `std` only, since the runtime should not use unsafe math anyway.

File: primitives/weights/src/weight_v2.rs
Comment (ggwpez) [inline primitives/weights/src/weight_v2.rs:416]: Is Rust not default implementing these?!

```suggestion
		*self = *self + other
```

File: primitives/weights/src/weight_v2.rs
Comment (ggwpez) [inline primitives/weights/src/weight_v2.rs:425]:
```suggestion
		*self = *self + other
```

**Review by ggwpez — APPROVED** [review]
Just ideas and nits.

**Review by nitwit69 — APPROVED** [review]
fmt lgtm

File: primitives/weights/src/weight_v2.rs
Comment (KiChjang) [reply primitives/weights/src/weight_v2.rs:129]: That also assumes that all fields are `u64`, which might not be the case in the future.

File: primitives/weights/src/weight_v2.rs
Comment (bkchr) [reply primitives/weights/src/weight_v2.rs:300]: Not really unsafe math.

**Review by bkchr — APPROVED** [review]

File: primitives/weights/src/weight_v2.rs
Comment (shawntabrizi) [reply primitives/weights/src/weight_v2.rs:39]: yes. but the whole PR will need migration, since weight is now 2-dimensional, so good to have both of these changes in at once.

File: primitives/weights/src/weight_v2.rs
Comment (shawntabrizi) [inline primitives/weights/src/weight_v2.rs:121]: this function will break if we add a third dimension in the future. does that imply we should not have it?

you could always do:

```
Weight::new().set_proof_size(a).set_ref_time(b)
```

Which would be equiv

File: primitives/weights/src/weight_v2.rs
Comment (shawntabrizi) [inline primitives/weights/src/weight_v2.rs:343]: we probably shouldnt impl `Mul` for weight. It makes no sense?

**Review by shawntabrizi — APPROVED** [review]
needs migration before we can merge right?

File: primitives/weights/src/weight_v2.rs
Comment (ggwpez) [reply primitives/weights/src/weight_v2.rs:129]: No... it just assumes that they are `Saturating`. This iter would probably not be a homogeneous collection but a custom type.

File: primitives/weights/src/weight_v2.rs
Comment (ggwpez) [reply primitives/weights/src/weight_v2.rs:121]: Well, it could be that it actually *should* break, to remind us to fix the code. Not sure…

Comment (KiChjang) [reply primitives/weights/src/weight_v2.rs:121]: That's what I was thinking as well, the `from_components` method should break whenever there's a new field. I've wondered whether we should just exclusively use a builder pattern, but having the option to use just 1 single function to create a `Weight` type sounded appealing to me.

Comment (shawntabrizi) [reply primitives/weights/src/weight_v2.rs]: i have no strong opinion in either direction. as long as we are being conscious of the choice i am happy :)

**Comment by KiChjang** [comment]
What I just realized is that we don't need a `CompactAs` implementation when the fields of the `Weight` struct are compact by default. In that case, when `Weight` appears in a parameter of an extrinsic, no `#[pallet::compact]` is necessary at all.

File: primitives/weights/src/weight_v2.rs
Comment (KiChjang) [reply primitives/weights/src/weight_v2.rs:136]: That, and the fact that all the fields must be homogeneous, because you're mapping over a collection of `Item`s, which must have the same type.

File: primitives/weights/src/weight_v2.rs
Comment (ggwpez) [reply primitives/weights/src/weight_v2.rs:45]: You could write a small test that encoded a WeightV1 and tries to decode it as new WeightV2. Probably need to account for the compact?

Comment (KiChjang) [reply primitives/weights/src/weight_v2.rs:45]: It is a bit difficult, since it requires a place where I can annotate `#[pallet::compact]`, and the only place I know is within the parameters of a dispatchable function. I'm not sure exactly how I can encode a `Weight` in the old way and try decoding it in the new way -- you can't really suddenly change the `Weight` parameter to a 2D weight type from a 1D one, yeah?

Comment (ggwpez) [reply primitives/weights/src/weight_v2.rs]: I am not sure if i really understood your pain-point, but is this not enough?  
```rust
#[test]
fn compact_weight_v1_can_be_decoded_as_v2() {
	use codec::{Decode, Encode, Compact};
	type WeightV1 = u64;

	let weight_v1: WeightV1 = 12345;
	let compact_weight_v1 = Compact(weight_v1);
	let encoded = Compact::<WeightV1>::encode(&compact_weight_v1);

	// Decode as weight v2
	let decoded: Weight = Decode::decode(&mut &encoded[..]).unwrap();
	assert_eq!(decoded.ref_time(), weight_v1);
}
```

Comment (KiChjang) [reply primitives/weights/src/weight_v2.rs]: It is not a bad unit test on its own, but it doesn't really test the `#[pallet::compact] OldWeight` to the `NewWeight` transition directly.

**Comment by KiChjang** [comment]
Okay so now CI is failing on runtime integrity tests, because the extrinsic weights have a proof size of 0, hence they are not strictly greater than 0, which is what we require:

```rust
// Make sure that if total is set it's greater than base_block &&
// base_for_class
error_assert!(
	(max_for_class.all_gt(self.base_block) && max_for_class.all_gt(base_for_class))
	|| max_for_class == Weight::zero(),
	&mut error,
	"[{:?}] {:?} (total) has to be greater than {:?} (base block) & {:?} (base extrinsic)",
	class, max_for_class, self.base_block, base_for_class,
);
// Max extrinsic can't be greater than max_for_class.
error_assert!(
	weights
		.max_extrinsic
		.unwrap_or(Weight::zero())
		.all_lte(max_for_class.saturating_sub(base_for_class)),
	&mut error,
	"[{:?}] {:?} (max_extrinsic) can't be greater than {:?} (max for class)",
	class,
	weights.max_extrinsic,
	max_for_class.saturating_sub(base_for_class),
);
// Max extrinsic should not be 0
error_assert!(
	weights.max_extrinsic.unwrap_or_else(Weight::max_value).all_gt(Weight::zero()),
	&mut error,
	"[{:?}] {:?} (max_extrinsic) must not be 0. Check base cost and average initialization cost.",
	class, weights.max_extrinsic,
);
// Make sure that if reserved is set it's greater than base_for_class.
error_assert!(
	reserved.all_gt(base_for_class) || reserved == Weight::zero(),
	&mut error,
	"[{:?}] {:?} (reserved) has to be greater than {:?} (base extrinsic) if set",
	class,
	reserved,
	base_for_class,
);
// Make sure max block is greater than max_total if it's set.
error_assert!(
	self.max_block.all_gte(weights.max_total.unwrap_or(Weight::zero())),
	&mut error,
	"[{:?}] {:?} (max block) has to be greater than {:?} (max for class)",
	class,
	self.max_block,
	weights.max_total,
);
// Make sure we can fit at least one extrinsic.
error_assert!(
	self.max_block.all_gt(base_for_class + self.base_block),
	&mut error,
	"[{:?}] {:?} (max block) must fit at least one extrinsic {:?} (base weight)",
	class,
	self.max_block,
	base_for_class + self.base_block,
);
```

Should these all be `any_*` instead of `all_*`? Or should we actually start filling up the proof size weight for each extrinsic?

**Comment by ggwpez** [comment]
> Or should we actually start filling up the proof size weight for each extrinsic?

I think we should, https://github.com/paritytech/substrate/issues/12251. But it could legitimately be that the `base_extrinsic` is zero, since we only read/write overlayed storage. So for now just short-circuit them with `any_gt` and add it as TODO to the issue?

**Comment by KiChjang** [comment]
There's a unit test that I keep failing right now, and it seems to be related to how the storage root has changed after executing a test extrinsic. Here's the code for the test:

```rust
fn block_import_works() {
	block_import_works_inner(
		new_test_ext_v0(1),
		hex!("1039e1a4bd0cf5deefe65f313577e70169c41c7773d6acf31ca8d671397559f5").into(),
	);
	block_import_works_inner(
		new_test_ext(1),
		hex!("75e7d8f360d375bbe91bcf8019c01ab6362448b4a89e3b329717eb9d910340e5").into(),
	);
}
fn block_import_works_inner(mut ext: sp_io::TestExternalities, state_root: H256) {
	ext.execute_with(|| {
		Executive::execute_block(Block {
			header: Header {
				parent_hash: [69u8; 32].into(),
				number: 1,
				state_root,
				extrinsics_root: hex!(
					"03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314"
				)
				.into(),
				digest: Digest { logs: vec![] },
			},
			extrinsics: vec![],
		});
	});
}
```

Am I correct in thinking that this is actually functioning as expected due to the addition of a new field in `Weight`, and we should instead change the expected state root for this test?

**Comment by cheme** [comment]
> Am I correct in thinking that this is actually functioning as expected due to the addition of a new field in `Weight`, and we should instead change the expected state root for this test?

Yes looks like weights are stored in state on initialization.
Changing the roots should be fine (
```
	#[test]
	fn block_import_works() {
		block_import_works_inner(
			new_test_ext_v0(1),
			hex!("0d786e24c1f9e6ce237806a22c005bbbc7dee4edd6692b6c5442843d164392de").into(),
		);
		block_import_works_inner(
			new_test_ext(1),
			hex!("348485a4ab856467b440167e45f99b491385e8528e09b0e51f85f814a3021c93").into(),
		);
	}
``` I think)

**Comment by KiChjang** [comment]
bot merge

**Comment by nazar-pc** [comment]
Hm... this is really annoying for chains that don't have PoV, now I need to learn the concept which is not applicable to the chain whatsoever. Is there a chance to make it optional somehow?

**Comment by KiChjang** [comment]
You can do so by setting the max proof size to `u64::MAX`. The docs somewhere should mention this, but unsure where exactly.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis/1026/2

**Pull request merged**
